File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.cpp

#define ll long long
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        sort(potions.begin(), potions.end());
        vector<int> ans;
        ll n = potions.size();
        for(auto spell: spells){
            ll mn = (success+spell-1)/spell;
            ll indx = lower_bound(potions.begin(), potions.end(), mn) - potions.begin();
            ans.push_back(n-indx);
        }
        return ans;
    }
};

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.java
   import java.util.*;
class Solution {


public int[] successfulPairs(int[] spells, int[] potions, long success) {
    Arrays.sort(potions);
    int n = spells.length;
    int m = potions.length;
    int[] ans = new int[n];

    for (int i = 0; i < n; i++) {
        long mn = (success + spells[i] - 1) / spells[i];
        int indx = binarySearch(potions, mn);
        ans[i] = m - indx;
    }

    return ans;
}

// Custom binary search to find the lower bound index
private int binarySearch(int[] potions, long value) {
    int left = 0, right = potions.length;
    while (left < right) {
        int mid = left + (right - left) / 2;
        if (potions[mid] < value) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}


}